modules/shared/src/main/scala/org/tessellation/currency/validations/DataTransactionsValidations.scala

```diff
@@ -0,0 +1,95 @@
+package org.tessellation.currency.validations
+
+import cats.data.{NonEmptyList, OptionT, ValidatedNec}
+import cats.effect.Async
+import cats.syntax.all._
+
+import org.tessellation.currency.dataApplication._
+import org.tessellation.currency.utils.DataTransactionUtils.{getDataUpdates, getFeeTransactionOfDataUpdate}
+import org.tessellation.currency.validations.FeeTransactionValidations.validateFeeTransaction
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Balance
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signed
+
+object DataTransactionsValidations {
+  private def validateDataTransactions[F[_]: Async: SecurityProvider](
+    dataTransactions: NonEmptyList[Signed[DataTransaction]],
+    balances: Map[Address, Balance],
+    dataApplication: BaseDataApplicationService[F],
+    validateFee: SnapshotOrdinal => (
+      Signed[DataUpdate],
+      Option[Signed[FeeTransaction]]
+    ) => F[ValidatedNec[DataApplicationValidationError, Unit]],
+    gsOrdinal: SnapshotOrdinal
+  ): F[ValidatedNec[DataApplicationValidationError, Unit]] = {
+
+    val dataUpdates = dataTransactions.collect {
+      case Signed(dataUpdate: DataUpdate, proofs) => Signed(dataUpdate, proofs)
+    }
+
+    if (dataUpdates.isEmpty) {
+      DataApplicationFeeError("Could not find any data update transaction")
+        .asInstanceOf[DataApplicationValidationError]
+        .invalidNec[Unit]
+        .pure[F]
+    } else {
+      NonEmptyList
+        .fromListUnsafe(dataUpdates)
+        .traverse { dataUpdate =>
+          for {
+            maybeFeeTransaction <- getFeeTransactionOfDataUpdate(dataTransactions, dataUpdate.value, dataApplication.serializeUpdate)
+            feeTransactionValidation <- validateFeeTransaction(maybeFeeTransaction, dataTransactions, balances, dataApplication)
+            feeAgainstDataUpdateValidation <- validateFee(gsOrdinal)(dataUpdate, maybeFeeTransaction)
+          } yield
+            feeTransactionValidation
+              .productR(feeAgainstDataUpdateValidation)
+        }
+        .map(_.reduce)
+    }
+
+  }
+
+  def validateDataTransactionsL1[F[_]: Async: L1NodeContext: SecurityProvider](
```

Comment:
is validation implementation different on separate layers? maybe you just need to extract a single fn for validation (I assume it has the same definition) rather than creating to separate validate fns with repeated functionality

Reply:
It's a little different because in L0 we have the validateData with the state, and in L1 we call the validateUpdate without state

Reply:
Ok, so it's a contextless and contextful validation

Reply:
@IPadawans bump